### ISSUE TITLE
Mutating Arrays not Allowed

### ISSUE THREAD

**Issue by CharlesRSmith44** [SciML/DiffEqFlux.jl#443]
Hello, I'm trying to write a program where for each instance of training I draw a new set of data. My NN takes in the set of observations drawn from a normal distribution and then estimates the normal distributions mean and variance. However, when I try to run the code, I get the error "Mutating Arrays not Allowed" when trying to train the NN. Any advice?

Code is below:

```
import Pkg; Pkg.add("IterTools")
using Random, Distributions, Flux, Plots
using Flux: @epochs
using IterTools: ncycle

# Setting the seed
Random.seed!(123) 

# Initial Inputs
μ = randn(1,1)[1] #mean
σ = randn(1,1)[1] #variance
σ = σ.^2 .+ 0.01
n_draws = 100 #number of draws 

# Drawing the dataset
ydata = [μ;σ]
NN = Chain(x -> x,
             Dense(n_draws,20,relu),
             Dense(20,20,relu),
             Dense(20,2))

function gen_data(μ,σ)
    xdata = rand(Normal(μ,σ),n_draws)
end 

function loss2(μ,σ) 
    xdata = gen_data(μ,σ)
    pred = NN(xdata)
    totalloss = sum((pred .- ydata).^2)
    return totalloss
end

@show loss2(μ,σ)
ps = Flux.params(NN)
evalcb() = @show(loss2(μ,σ))

### Training

# train model on the same data num_cycle times
num_cycles = 1
data = ncycle([(μ,σ)], num_cycles)

# trainign the model 
opt = ADAM(0.01)
@epochs 100 Flux.train!(loss2, ps, data, opt, cb=evalcb)
```

**Comment by ali-ramadhan** [comment]
Always helpful if you can post the stacktrace/error output so it's easier to debug/help. Also posting your Julia environment (output of `] status` at the REPL) helps.

I think the default auto-diff used by DiffEqFlux.jl is Zygote.jl which does not generally allow for arrays to be mutated. Can't see exactly where your code is mutating an error though (which is where a stacktrace would be helpful).

**Comment by smitch151** [comment]
The error is this:

Mutating arrays is not supported

Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] (::Zygote.var"#455#456")(::Nothing) at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/lib/array.jl:64
 [3] (::Zygote.var"#999#back#457"{Zygote.var"#455#456"})(::Nothing) at /home/ec2-user/.julia/packages/ZygoteRules/6nssF/src/adjoint.jl:49
 [4] rand! at /home/ec2-user/.julia/packages/Distributions/jFoHB/src/univariates.jl:165 [inlined]
 [5] (::typeof(∂(rand!)))(::Array{Float64,1}) at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/compiler/interface2.jl:0
 [6] rand at /home/ec2-user/.julia/packages/Distributions/jFoHB/src/univariates.jl:158 [inlined]
 [7] (::typeof(∂(rand)))(::Array{Float64,1}) at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/compiler/interface2.jl:0
 [8] rand at /home/ec2-user/.julia/packages/Distributions/jFoHB/src/genericrand.jl:24 [inlined]
 [9] gen_data at ./In[3]:16 [inlined]
 [10] (::typeof(∂(gen_data)))(::Array{Float64,1}) at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/compiler/interface2.jl:0
 [11] loss2 at ./In[3]:20 [inlined]
 [12] (::typeof(∂(loss2)))(::Float64) at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/compiler/interface2.jl:0
 [13] #175 at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/lib/lib.jl:182 [inlined]
 [14] #347#back at /home/ec2-user/.julia/packages/ZygoteRules/6nssF/src/adjoint.jl:49 [inlined]
 [15] #17 at /home/ec2-user/.julia/packages/Flux/Fj3bt/src/optimise/train.jl:89 [inlined]
 [16] (::typeof(∂(λ)))(::Float64) at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/compiler/interface2.jl:0
 [17] (::Zygote.var"#50#51"{Zygote.Params,Zygote.Context,typeof(∂(λ))})(::Float64) at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/compiler/interface.jl:177
 [18] gradient(::Function, ::Zygote.Params) at /home/ec2-user/.julia/packages/Zygote/1GXzF/src/compiler/interface.jl:54
 [19] macro expansion at /home/ec2-user/.julia/packages/Flux/Fj3bt/src/optimise/train.jl:88 [inlined]
 [20] macro expansion at /home/ec2-user/.julia/packages/Juno/n6wyj/src/progress.jl:134 [inlined]
 [21] train!(::typeof(loss2), ::Zygote.Params, ::IterTools.NCycle{Array{Tuple{Float64,Float64},1}}, ::ADAM; cb::typeof(evalcb)) at /home/ec2-user/.julia/packages/Flux/Fj3bt/src/optimise/train.jl:81
 [22] top-level scope at /home/ec2-user/.julia/packages/Flux/Fj3bt/src/optimise/train.jl:122
 [23] top-level scope at /home/ec2-user/.julia/packages/Juno/n6wyj/src/progress.jl:134
 [24] top-level scope at In[3]:38

My Repl is this: 
Status `~/.julia/environments/v1.4/Project.toml`
  [7d9fca2a] Arpack v0.4.0
  [6e4b80f9] BenchmarkTools v0.5.0
  [2d3116d5] Bridge v0.11.4
  [336ed68f] CSV v0.7.7
  [052768ef] CUDA v0.1.0
  [be33ccc6] CUDAnative v3.0.4
  [3a865a2d] CuArrays v2.2.2
  [6cf23378] DSGE v1.1.6
  [a93c6f00] DataFrames v0.21.8
  [aae7a2af] DiffEqFlux v1.17.0
  [9fdde737] DiffEqOperators v4.12.0
  [41bf760c] DiffEqSensitivity v6.33.0
  [0c46a032] DifferentialEquations v6.15.0
  [31c24e10] Distributions v0.23.8
  [a3315474] EconPDEs v0.4.1
  [2fe49d83] Expectations v1.6.0
  [587475ba] Flux v0.10.4
  [6e85b7ee] FredData v0.4.0
  [0c68f7d7] GPUArrays v3.4.1
  [28b8d3ca] GR v0.51.0
  [7073ff75] IJulia v1.22.0
  [2fce0c6f] InfinitesimalGenerators v0.3.4
  [c8e1da08] IterTools v1.3.0
  [442fdcdd] Measures v0.3.1
  [e47e5152] ModelConstructors v0.1.12
  [2774e3e8] NLsolve v4.4.1
  [8faf48c0] NeuralNetDiffEq v1.6.0
  [0af1e411] Opus v1.0.0
  [91a5bcdd] Plots v1.6.12
  [fcd29c91] QuantEcon v0.16.2
  [79098fc4] Rmath v0.6.1
  [c6043b28] SMC v0.1.11
  [00bf1f32] SolveDSGE v0.3.11
  [261fd431] StateSpaceRoutines v0.3.3
  [f3b207a7] StatsPlots v0.14.15
  [789caeaf] StochasticDiffEq v6.25.0
  [34f15cae] StructJuMP v0.2.0
  [c3572dad] Sundials v4.2.6
  [bd369af6] Tables v1.1.0
  [9f7883ad] Tracker v0.2.12
  [e88e6eb3] Zygote v0.4.22
  [8ba89e20] Distributed 
  [37e2e46d] LinearAlgebra 
  [10745b16] Statistics 
  [8dfed614] Test

**Comment by ali-ramadhan** [comment]
Seems that Distributions.jl mutates arrays when generating values in `rand` which would make Zygote.jl unhappy. Not familiar with Distributions.jl but maybe it has a method of pulling random values from a distribution without mutating arrays?

If you just need to pull from a Gaussian distribution you could probably use `randn`?

**Comment by ChrisRackauckas** [comment]
You might need to `using DistributionsAD`

**Comment by smitch151** [comment]
Okay, so I got it to work using randn and DistributionsAD. However, I think randn restricts to mean = 0, var = 1. Do I just have to rescale the data given a randn draw to get mean = m, var =v?

**Comment by ChrisRackauckas** [comment]
`sqrt(v)*randn() + m`

**Comment by Emmanuel-R8** [comment]
I am jumping in with a question having made no research whatsoever.

Is the reason to restrict rand() to m=0;var=1 related to the variational autoencoders' reparametrisation trick?

**Comment by ali-ramadhan** [comment]
In Base Julia `rand` is a pretty general function for picking a random value from a set of values (see https://docs.julialang.org/en/v1/stdlib/Random/#Base.rand) that happens to default to `rand(Float64)` which picks a random float in the range [0, 1). So I don't think there's any restriction.

I guess you can always scale the output of `rand` and `randn` to obtain the uniform and Gaussian distributions you want.

I think `rand` and `randn` were implemented to be as general purpose as possible.

**Comment by ChrisRackauckas** [comment]
This package directly `using DistributionsAD` which alleviates this issue.
